src/main/scala/is/hail/utils/richUtils/RichDenseMatrixDouble.scala

```diff
@@ -1,62 +1,27 @@
 package is.hail.utils.richUtils
 
-import java.io.{DataInputStream, DataOutputStream}
+import java.io.{InputStream, OutputStream}
 
 import breeze.linalg.DenseMatrix
 import is.hail.HailContext
-import is.hail.annotations.Memory
 import is.hail.distributedmatrix.{BlockMatrix, BlockMatrixMetadata, GridPartitioner}
+import is.hail.io.CodecSpec
 import is.hail.utils._
-import org.apache.commons.lang3.StringUtils
 import org.json4s.jackson
 
 object RichDenseMatrixDouble {
-  // copies n doubles as bytes from data to dos
-  def writeDoubles(dos: DataOutputStream, data: Array[Double], n: Int) {
-    assert(n <= data.length)
-    var nLeft = n
-    val bufSize = math.min(nLeft, 8192)
-    val buf = new Array[Byte](bufSize << 3) // up to 64KB of doubles
-
-    var off = 0
-    while (nLeft > 0) {
-      val nCopy = math.min(nLeft, bufSize)
-
-      Memory.memcpy(buf, 0, data, off, nCopy)
-      dos.write(buf, 0, nCopy << 3)
-
-      off += nCopy
-      nLeft -= nCopy
-    }
-  }
-
-  // copies n doubles as bytes from dis to data
-  def readDoubles(dis: DataInputStream, data: Array[Double], n: Int) {
-    assert(n <= data.length)
-    var nLeft = n
-    val bufSize = math.min(nLeft, 8192)
-    val buf = new Array[Byte](bufSize << 3) // up to 64KB of doubles
-
-    var off = 0
-    while (nLeft > 0) {
-      val nCopy = math.min(nLeft, bufSize)
-
-      dis.readFully(buf, 0, nCopy << 3)
-      Memory.memcpy(data, off, buf, 0, nCopy)
-
-      off += nCopy
-      nLeft -= nCopy
-    }
-  }
-
-  // assumes zero offset and minimal majorStride 
-  def read(dis: DataInputStream): DenseMatrix[Double] = {
-    val rows = dis.readInt()
-    val cols = dis.readInt()
-    val isTranspose = dis.readBoolean()
+  val bufferSpec = CodecSpec.blockSpecs(1)
```

Comment:
Don't use `blockSpecs`, those are really just for testing (but used for python so I couldn't put them in the tests area).

The buffer spec needs to be written to the metadata and that's the spec you need to use on reading.  That lets you change the buffer spec in a forward compatible way (e.g. add a double dictionary encoding buffer as Richard suggested).  See e.g. Table.read/write for examples.

Reply:
Also, slightly strange place to put this.  It should originate in the write function and be passed around from there.